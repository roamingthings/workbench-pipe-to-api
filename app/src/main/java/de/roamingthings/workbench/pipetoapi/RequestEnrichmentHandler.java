package de.roamingthings.workbench.pipetoapi;

import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import io.micronaut.function.aws.MicronautRequestHandler;
import io.micronaut.json.JsonMapper;
import jakarta.inject.Inject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class RequestEnrichmentHandler extends MicronautRequestHandler<List<SQSEvent.SQSMessage>, List<ApiDestinationRequest>> {

    private static final Logger log = LoggerFactory.getLogger(RequestEnrichmentHandler.class);

    @Inject
    JsonMapper objectMapper;

    @Override
    public List<ApiDestinationRequest> execute(List<SQSEvent.SQSMessage> input) {
        log.info("Enriching {} messages", input.size());
        return input.parallelStream()
                .map(SQSEvent.SQSMessage::getBody)
                .map(ApiDestinationRequest::new)
                .toList();
    }
}
