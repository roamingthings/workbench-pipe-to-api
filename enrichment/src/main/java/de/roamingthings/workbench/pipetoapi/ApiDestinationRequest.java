package de.roamingthings.workbench.pipetoapi;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.micronaut.serde.annotation.Serdeable;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;

@Serdeable
public record ApiDestinationRequest(@JsonProperty("Headers") Map<String, String> headers, @JsonProperty("RequestBody") String requestBody) {
    public ApiDestinationRequest(String body) {
        this(calcSignature(body), body);
    }

    private static Map<String, String> calcSignature(String body) {
        try {
            var md = MessageDigest.getInstance("SHA-1");
            var digest = md.digest(body.getBytes());
            var sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            var headers = new HashMap<String, String>();
            headers.put("signature", sb.toString());
            return headers;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-1 algorithm not found", e);
        }
    }
}
