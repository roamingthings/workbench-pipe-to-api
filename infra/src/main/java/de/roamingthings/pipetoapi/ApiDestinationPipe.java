package de.roamingthings.pipetoapi;

import lombok.Builder;
import software.amazon.awscdk.RemovalPolicy;
import software.amazon.awscdk.services.events.IApiDestination;
import software.amazon.awscdk.services.iam.Effect;
import software.amazon.awscdk.services.iam.PolicyDocument;
import software.amazon.awscdk.services.iam.PolicyStatement;
import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.iam.ServicePrincipal;
import software.amazon.awscdk.services.lambda.IFunction;
import software.amazon.awscdk.services.logs.ILogGroup;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.logs.RetentionDays;
import software.amazon.awscdk.services.pipes.CfnPipe;
import software.amazon.awscdk.services.sqs.IQueue;
import software.constructs.Construct;

import java.util.List;
import java.util.Map;

public class ApiDestinationPipe extends Construct {

    private static final Number DEFAULT_BATCH_SIZE = 1;
    private static final Number DEFAULT_BATCHING_WINDOW_SECONDS = 6;

    public ApiDestinationPipe(Construct scope, String id, ApiDestinationPipeProps props) {
        super(scope, id);

        var logGroup = createLogGroup(getNode().getId());
        createPipe(props.sourceQueue, props.enrichmentFunction, props.apiDestinationTarget, logGroup);
    }

    private ILogGroup createLogGroup(String name) {
        return LogGroup.Builder.create(this, "LogGroup")
                .logGroupName("/aws/vendedlogs/pipes/%s".formatted(name))
                .retention(RetentionDays.ONE_DAY)
                .removalPolicy(RemovalPolicy.DESTROY)
                .build();
    }

    private Role createPipeRole(IQueue sourceQueue, IFunction enrichmentFunction, IApiDestination apiDestinationTarget, ILogGroup logGroup) {
        var role = Role.Builder.create(this, "Role")
                .inlinePolicies(Map.of(
                        "targetPolicy", createInvokeApiDestinationPolicy(apiDestinationTarget)
                ))
                .assumedBy(new ServicePrincipal("pipes.amazonaws.com"))
                .build();
        sourceQueue.grantConsumeMessages(role);
        enrichmentFunction.grantInvoke(role);
        logGroup.grantWrite(role);
        return role;
    }

    private static PolicyDocument createInvokeApiDestinationPolicy(IApiDestination apiDestinationTarget) {
        var apiDestinationArn = apiDestinationTarget.getApiDestinationArn();
        return PolicyDocument.Builder.create()
                .statements(List.of(
                        PolicyStatement.Builder.create()
                                .effect(Effect.ALLOW)
                                .actions(List.of("events:InvokeApiDestination"))
                                .resources(List.of(apiDestinationArn))
                                .build()
                ))
                .build();
    }

    private void createPipe(IQueue source, IFunction enrichment, IApiDestination target, ILogGroup logGroup) {
        var role = createPipeRole(source, enrichment, target, logGroup);
        CfnPipe.Builder.create(this, "Pipe")
                .source(source.getQueueArn())
                .sourceParameters(CfnPipe.PipeSourceParametersProperty.builder()
                        .sqsQueueParameters(CfnPipe.PipeSourceSqsQueueParametersProperty.builder()
                                .batchSize(DEFAULT_BATCH_SIZE)
                                .maximumBatchingWindowInSeconds(DEFAULT_BATCHING_WINDOW_SECONDS)
                                .build())
                        .build()
                )
                .enrichment(enrichment.getFunctionArn())
                .enrichmentParameters(CfnPipe.PipeEnrichmentParametersProperty.builder()
//                        .inputTemplate("${body})
                        .build())
                .target(target.getApiDestinationArn())
                .targetParameters(CfnPipe.PipeTargetParametersProperty.builder()
                        .httpParameters(CfnPipe.PipeTargetHttpParametersProperty.builder()
                                .headerParameters(Map.of(
                                        "X-Signature", "$.Headers.signature"
                                ))
                                .build())
                        .inputTemplate("<$.RequestBody>")
                        .build())
                .roleArn(role.getRoleArn())
                // See https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_PipeLogConfiguration.html for
                // available logging configuration options
                // Available levels: OFF, ERROR, INFO, TRACE
                .logConfiguration(CfnPipe.PipeLogConfigurationProperty.builder()
                        .cloudwatchLogsLogDestination(CfnPipe.CloudwatchLogsLogDestinationProperty.builder()
                                .logGroupArn(logGroup.getLogGroupArn())
                                .build())
                        .level("TRACE")
                        .includeExecutionData(List.of("ALL"))
                        .build())
                .build();
    }

    @Builder
    public static class ApiDestinationPipeProps {
        private String name;
        private IQueue sourceQueue;
        private IFunction enrichmentFunction;
        private IApiDestination apiDestinationTarget;
    }
}
