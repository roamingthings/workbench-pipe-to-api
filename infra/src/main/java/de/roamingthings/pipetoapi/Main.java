package de.roamingthings.pipetoapi;

import software.amazon.awscdk.App;

public class Main {
    public static void main(final String[] args) {
        App app = new App();
        new PipeToApiStack(app, "Workbench-PipeToApi-Stack", PipeToApiStack.PipeToApiStackProps.builder()
                .endpointUrl(endointUrl(app))
                .build());
        app.synth();
    }

    protected static String endointUrl(App app) {
        Object endpointUrl = app.getNode().tryGetContext("endpointUrl");
        if (!(endpointUrl instanceof String)) {
            System.err.println("""
                    You need to specify an endpoint URL context variable when deploying this stack.
                    For example if your endpoint is https://api.example.com, you would run:
                    cdk deploy -c endpointUrl=https://api.example.com
                    """.stripIndent());
            System.exit(1);
        }
        return (String) endpointUrl;
    }
}