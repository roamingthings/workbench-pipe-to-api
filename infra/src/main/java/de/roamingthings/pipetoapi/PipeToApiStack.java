package de.roamingthings.pipetoapi;

import io.micronaut.aws.cdk.function.MicronautFunction;
import io.micronaut.aws.cdk.function.MicronautFunctionFile;
import io.micronaut.starter.application.ApplicationType;
import io.micronaut.starter.options.BuildTool;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.Environment;
import software.amazon.awscdk.IStackSynthesizer;
import software.amazon.awscdk.PermissionsBoundary;
import software.amazon.awscdk.SecretValue;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.events.ApiDestination;
import software.amazon.awscdk.services.events.Authorization;
import software.amazon.awscdk.services.events.Connection;
import software.amazon.awscdk.services.events.HttpMethod;
import software.amazon.awscdk.services.lambda.Alias;
import software.amazon.awscdk.services.lambda.Architecture;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.Runtime;
import software.amazon.awscdk.services.lambda.SnapStartConf;
import software.amazon.awscdk.services.lambda.Tracing;
import software.amazon.awscdk.services.logs.RetentionDays;
import software.amazon.awscdk.services.sqs.Queue;
import software.constructs.Construct;

import java.util.HashMap;
import java.util.Map;

public class PipeToApiStack extends Stack {

    public static final String UNSAFE_DEMONSTRATION_API_KEY = "my-api-key";

    public PipeToApiStack(Construct parent, String id, PipeToApiStackProps props) {
        super(parent, id, props);

        var sourceQueue = Queue.Builder.create(this, "Source")
                .build();

        var environmentVariables = new HashMap<String, String>();
        var function = MicronautFunction.create(ApplicationType.FUNCTION,
                false,
                this,
                "RequestEnrichment")
                .runtime(Runtime.JAVA_21)
                .handler("de.roamingthings.workbench.pipetoapi.RequestEnrichmentHandler")
                .environment(environmentVariables)
                .code(Code.fromAsset(functionPath()))
                .timeout(Duration.seconds(10))
                .memorySize(768)
                .logRetention(RetentionDays.ONE_DAY)
                .logFormat("JSON")
                .tracing(Tracing.ACTIVE)
                .architecture(Architecture.X86_64)
                .snapStart(SnapStartConf.ON_PUBLISHED_VERSIONS)
                .build();
        var alias = Alias.Builder.create(this, "Alias")
                .aliasName("live")
                .version(function.getCurrentVersion())
                .build();

        var apiDestinationTarget = createApiDestinationTarget(
                props.endpointUrl,
                Authorization.apiKey("Authorization", SecretValue.unsafePlainText(UNSAFE_DEMONSTRATION_API_KEY))
        );

        new ApiDestinationPipe(this, "ApiDestinationPipe", ApiDestinationPipe.ApiDestinationPipeProps.builder()
                .sourceQueue(sourceQueue)
                .enrichmentFunction(alias)
                .apiDestinationTarget(apiDestinationTarget)
                .build());
    }

    public static String functionPath() {
        return "../app/build/libs/" + functionFilename();
    }

    public static String functionFilename() {
        return MicronautFunctionFile.builder()
            .graalVMNative(false)
            .version("0.1")
            .archiveBaseName("app")
            .buildTool(BuildTool.GRADLE)
            .build();
    }

    private ApiDestination createApiDestinationTarget(String endpointUrl, Authorization authorization) {
         var connection = Connection.Builder.create(this, "Connection")
                 .authorization(authorization)
                 .description("Connection with 3rd party API")
                 .build();

         return ApiDestination.Builder.create(this, "ApiDestination")
                 .endpoint(endpointUrl)
                 .httpMethod(HttpMethod.POST)
                 .rateLimitPerSecond(5)
                 .connection(connection)
                 .build();
     }

     @lombok.Builder
     public static class PipeToApiStackProps implements StackProps {
        private final Boolean analyticsReporting;
        private final Boolean crossRegionReferences;
        private final String description;
        private final Environment env;
        private final PermissionsBoundary permissionsBoundary;
        private final String stackName;
        private final Boolean suppressTemplateIndentation;
        private final IStackSynthesizer synthesizer;
        private final Map<String, String> tags;
        private final Boolean terminationProtection;

         private final String endpointUrl;
     }
}