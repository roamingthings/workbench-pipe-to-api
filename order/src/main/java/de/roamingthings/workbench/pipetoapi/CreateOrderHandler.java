package de.roamingthings.workbench.pipetoapi;

import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPResponse;
import io.micronaut.function.aws.MicronautRequestHandler;
import io.micronaut.json.JsonMapper;
import jakarta.inject.Inject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CreateOrderHandler extends MicronautRequestHandler<APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse> {

    private static final Logger log = LoggerFactory.getLogger(CreateOrderHandler.class);

    @Inject
    JsonMapper objectMapper;

    @Override
    public APIGatewayV2HTTPResponse execute(APIGatewayV2HTTPEvent input) {
        log.info("Received request: {}", input);
        return APIGatewayV2HTTPResponse.builder()
                .withStatusCode(200)
                .build();
    }
}
