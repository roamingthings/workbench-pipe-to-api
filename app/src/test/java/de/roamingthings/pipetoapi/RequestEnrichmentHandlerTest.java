package de.roamingthings.pipetoapi;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import de.roamingthings.workbench.pipetoapi.RequestEnrichmentHandler;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.List;

@ExtendWith(SnapshotExtension.class)
class RequestEnrichmentHandlerTest {

    static RequestEnrichmentHandler handler;

    Expect expect;

    @BeforeAll
    static void setupServer() {
        handler = new RequestEnrichmentHandler();
    }

    @AfterAll
    static void stopServer() {
        if (handler != null) {
            handler.getApplicationContext().close();
        }
    }

    @Test
    void testHandler() {
        var sqsMessage = new SQSEvent.SQSMessage();
        sqsMessage.setBody("""
            {
              "Foo": "Bar"
            }
        """.trim());
        var input = List.of(
                sqsMessage
        );

        var response = handler.execute(input);

        expect.serializer("json").toMatchSnapshot(response);
    }
}
