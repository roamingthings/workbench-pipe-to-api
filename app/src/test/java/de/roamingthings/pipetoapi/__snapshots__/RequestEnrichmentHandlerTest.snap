de.roamingthings.pipetoapi.RequestEnrichmentHandlerTest.testHandler=[
  [
    {
      "Headers": {
        "signature": "4d640ceb9523875b6f3940a7a516493274b304f1"
      },
      "RequestBody": "{\n      \"Foo\": \"Bar\"\n    }"
    }
  ]
]